envoy/matcher/matcher.h

```diff
@@ -25,7 +25,25 @@ class ServerFactoryContext;
 
 namespace Matcher {
 
-using MatchingDataType = absl::variant<absl::monostate, std::string>;
+// Abstract interface for custom matching data.
+// Overrides this interface to provide custom matcher specific implementation.
+struct CustomMatchData {
+  CustomMatchData() = default;
+
+  // Non-copyable/non-assignable
+  CustomMatchData(const CustomMatchData&) = delete;
+  CustomMatchData& operator=(const CustomMatchData&) = delete;
+
+  // Move-constructible/move-assignable
+  CustomMatchData(CustomMatchData&& other) = default;
+  CustomMatchData& operator=(CustomMatchData&& other) = default;
```

Comment:
why these are necessary? 🤔

Reply:
You generally work with interfaces as pointers or references, where move and assign semantics for the objects do not matter. I suggest removing all the defaulted constructors and just provide the default virtual destructor.

Reply:
Thanks! Done.